docs/csharp/nullable-references.md

```diff
@@ -0,0 +1,74 @@
+---
+title: Nullable reference types
+description: This article provides an overview of nullable reference types, added in C# 8. You'll learn how the feature provides safety against null reference exceptions, for new and existing projects.
+ms.date: 12/03/2018
+---
+# Nullable reference types
+
+C# 8 introduces **nullable reference types** and **non-nullable reference types** that enable you to make important statements about the properties for reference type variables:
+
+- **A reference is not supposed to be null**. When variables aren't supposed to be null, the compiler enforces these rules: The variable must be initialized to a non-null value. The variable can never be assigned the value `null`. Because of those two rules, it's safe to de-reference the variable without first checking that it isn't null.
+- **A reference may be null**. When variables may be null, the compiler enforces different rules. The variable may only be de-referenced when the compiler can guarantee that the value isn't null. These variables may be initialized with the default `null` value, and may be assigned the value `null` in other code.
+
+This new feature provides significant benefits over the earlier reference variables where the design intent couldn't be determined from the variable declaration. The compiler didn't provide safety for reference types:
+
+- **A reference could be null**. No warnings are issued when a reference type is initialized to null, or later assigned to null.
+- **A reference is assumed to be not null**. The compiler doesn't issue any warnings when reference types are dereferenced.
+
+A **nullable reference type** is noted using the same syntax as [nullable value types](programming-guide/nullable-types/index.md): a `?` is appended to the type of the variable. For example, the following variable declaration represents a nullable string variable, `name`:
+
+```csharp
+string? name;
+```
+
+Any variable where the `?` is not appended to the type name is a **non-nullable reference type**. That includes all reference type variables in existing code, when you have enabled this feature.
+
+The compiler uses static analysis to determine if a nullable reference is known to be non-null. THe compiler warns you if you dereference a nullable reference when it may be null. You can override this behavior using the **null-forgiving operator** (`!`) following a variable name. For example, if you know the `name` variable isn't null but the compiler issues a warning, you can write the following code to override the compiler's analysis:
```

Comment:
> should we add a link to the operators if we have those docs?

It's not yet documented. In the language reference, I would mention the name of the operator with a brief example and then reference this article. Because one needs to understand the whole feature to make sense of the operator. The language reference page would catch the folks searching for the meaning of `!` and redirect here. If that sounds like a plan, I'd be happy to update the following article: https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/logical-negation-operator